pkg/util/tas/tas_assignment.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tas
+
+import (
+	"iter"
+	"slices"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta2"
+)
+
+type TopologyAssignment struct {
+	Levels  []string
+	Domains []TopologyDomainAssignment
+}
+
+type TopologyDomainAssignment struct {
+	Values []string
+	Count  int32
+}
+
+func valueAtIndex(values kueue.TopologyAssignmentSliceLevelValues, idx int) string {
+	if values.Roots == nil {
+		return *values.UniversalValue
+	}
+	var prefix string
+	if values.Prefix != nil {
+		prefix = *values.Prefix
+	}
+	var suffix string
+	if values.Suffix != nil {
+		suffix = *values.Suffix
+	}
+	return prefix + values.Roots[idx] + suffix
+}
+
+func countAtIndex(slice kueue.TopologyAssignmentSlice, idx int) int32 {
+	if slice.PodCounts == nil {
+		return *slice.UniversalPodCount
+	}
+	return slice.PodCounts[idx]
+}
+
+func ValuesAtLevel(ta *kueue.TopologyAssignment, levelIdx int) iter.Seq[string] {
+	return func(yield func(string) bool) {
+		for _, slice := range ta.Slices {
+			values := slice.ValuesPerLevel[levelIdx]
+			for i := range int(slice.DomainCount) {
+				if !yield(valueAtIndex(values, i)) {
+					return
+				}
+			}
+		}
+	}
+}
+
+func PodCounts(ta *kueue.TopologyAssignment) iter.Seq[int32] {
+	return func(yield func(int32) bool) {
+		for _, slice := range ta.Slices {
+			for i := range int(slice.DomainCount) {
+				if !yield(countAtIndex(slice, i)) {
+					return
+				}
+			}
+		}
+	}
+}
+
+func TotalDomainCount(a *kueue.TopologyAssignment) int {
+	res := 0
+	for _, slice := range a.Slices {
+		res += int(slice.DomainCount)
+	}
+	return res
+}
+
+func InternalSeqFrom(ta *kueue.TopologyAssignment) iter.Seq[TopologyDomainAssignment] {
+	if ta == nil {
+		return nil
+	}
+	return func(yield func(TopologyDomainAssignment) bool) {
+		for _, slice := range ta.Slices {
+			for i := range int(slice.DomainCount) {
+				req := TopologyDomainAssignment{
+					Count:  countAtIndex(slice, i),
```

Comment:
Nit: Looking at the `countAtIndex` implementation:
```go
func countAtIndex(slice kueue.TopologyAssignmentSlice, idx int) int32 {
	if slice.PodCounts == nil {
		return *slice.UniversalPodCount
	}
	return slice.PodCounts[idx]
}
```

Won't this make `slice.DomainCount` checks of `slice.PodCounts == nil` when it could have done a single one? Same in the `PodCounts` function.

Quickly looking at the definitions, do I see correctly that there might potentially be `100000` of checks? I'm not saying this is a big deal. Maybe there should be an iterator like `PodCounts` but for `TopologyAssignmentSlice` or smth?